File: modules/wem-api/src/main/java/com/enonic/wem/query/expr/LogicalExpr.java
package com.enonic.wem.query.expr;

public final class LogicalExpr
    implements ConstraintExpr
{
    public enum Operator
    {
        AND,
        OR
    }

    private final ConstraintExpr left;

    private final ConstraintExpr right;

    private final Operator operator;

    private LogicalExpr( final ConstraintExpr left, final Operator operator, final ConstraintExpr right )
    {
        this.left = left;
        this.right = right;
        this.operator = operator;
    }

    public ConstraintExpr getLeft()
    {
        return this.left;
    }

    public ConstraintExpr getRight()
    {
        return this.right;
    }

    public Operator getOperator()
    {
        return this.operator;
    }

    @Override
    public String toString()
    {
        return "(" + this.left.toString() + " " + this.operator.toString() + " " + this.right.toString() + ")";
    }

    public static LogicalExpr and( final ConstraintExpr left, final ConstraintExpr right )
    {
        return new LogicalExpr( left, Operator.AND, right );
    }

    public static LogicalExpr or( final ConstraintExpr left, final ConstraintExpr right )
    {
        return new LogicalExpr( left, Operator.OR, right );
    }
}

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/query/expr/LogicalExpr.ts
module api.query.expr {

    export class LogicalExpr implements ConstraintExpr {

        private left: ConstraintExpr;
        private right: ConstraintExpr;
        private operator: LogicalOperator;

        constructor(left: ConstraintExpr, operator: LogicalOperator, right: ConstraintExpr) {
            this.left = left;
            this.right = right;
            this.operator = operator;
        }

        getLeft(): ConstraintExpr {
            return this.left;
        }

        getRight(): ConstraintExpr {
            return this.right;
        }

        getOperator(): LogicalOperator {
            return this.operator;
        }

        toString() {
            return "(" + this.left.toString() + " " + this.operatorAsString() + " " + this.right.toString() + ")";
        }

        public static and(left: ConstraintExpr, right: ConstraintExpr): LogicalExpr {
            return new LogicalExpr(left, LogicalOperator.AND, right);
        }

        public static or(left: ConstraintExpr, right: ConstraintExpr): LogicalExpr {
            return new LogicalExpr(left, LogicalOperator.OR, right);
        }

        private operatorAsString(): string {
            switch (this.operator) {
            case LogicalOperator.AND:
                return "AND";
            case LogicalOperator.OR:
                return "OR";
            default:
                return "";
            }
        }
    }
}
